destroy() of permanent entity after failover crashes server instead of throwing PermanentEntityException

To Reproduce:

1) permanent entity instance is created while Active-Passive cluster is booted.
2) active server is stopped and mirror  failover
3) client calls EntityRef.destroy() on permanent entity.
Expected result: PermanentEntityException is thrown
Actual result: server is crashed with the following error on server:

com.tc.util.TCAssertionError: Assertion failed
at com.tc.util.Assert.failure(Assert.java:55)
at com.tc.util.Assert.eval(Assert.java:64)
at com.tc.util.Assert.assertTrue(Assert.java:85)
at com.tc.objectserver.entity.ManagedEntityImpl.destroyEntity(ManagedEntityImpl.java:600)
at com.tc.objectserver.entity.ManagedEntityImpl.invokeLifecycleOperation(ManagedEntityImpl.java:424)
at com.tc.objectserver.entity.ManagedEntityImpl.lambda$processLifecycleEntity$0(ManagedEntityImpl.java:255)
at com.tc.objectserver.entity.ManagedEntityImpl$SchedulingRunnable.run(ManagedEntityImpl.java:1048)
at com.tc.objectserver.entity.RequestProcessor$EntityRequest.invoke(RequestProcessor.java:170)
at com.tc.objectserver.entity.RequestProcessor$EntityRequest.run(RequestProcessor.java:161)
at com.tc.objectserver.entity.RequestProcessorHandler.handleEvent(RequestProcessorHandler.java:27)
at com.tc.objectserver.entity.RequestProcessorHandler.handleEvent(RequestProcessorHandler.java:23)
at com.tc.async.impl.StageQueueImpl$FlushingHandledContext.runWithHandler(StageQueueImpl.java:527)
at com.tc.async.impl.StageImpl$WorkerThread.run(StageImpl.java:192)

--- a/dso-l2/src/main/java/com/tc/objectserver/entity/ManagedEntityImpl.java
+++ b/dso-l2/src/main/java/com/tc/objectserver/entity/ManagedEntityImpl.java
@@ -842,7 +842,11 @@ public class ManagedEntityImpl implements ManagedEntity {
   
   @Override
   public void resetReferences(int count) {
-    this.clientReferenceCount = count;
+    if (canDelete) {
+      this.clientReferenceCount = count;
+    } else {
+      Assert.assertEquals(this.clientReferenceCount, ManagedEntityImpl.UNDELETABLE_ENTITY);
+    }
   }
   
   @Override
@@ -854,8 +858,11 @@ public class ManagedEntityImpl implements ManagedEntity {
     this.isInActiveState = true;
 // any clients, previously connected will reconnect or not.  Failed reconnects will be cleaned
 // up on passives 
-    this.clientReferenceCount = 0;
-
+    if (canDelete) {
+      this.clientReferenceCount = 0;
+    } else {
+      Assert.assertEquals(this.clientReferenceCount, ManagedEntityImpl.UNDELETABLE_ENTITY);
+    }
     if (!this.isDestroyed) {
       logger.info("promoteEntity: " + getID());
       if (null != this.passiveServerEntity) {

--- a/dso-l2/src/test/java/com/tc/objectserver/entity/ManagedEntityImplTest.java
+++ b/dso-l2/src/test/java/com/tc/objectserver/entity/ManagedEntityImplTest.java
@@ -253,6 +253,27 @@ public class ManagedEntityImplTest {
   
   
   @Test 
+  public void testPromotePermanentEntity() throws Exception {
+    // first create a passive entity and then promote
+    managedEntity = new ManagedEntityImpl(entityID, version, consumerID, loopback, serviceRegistry, clientEntityStateManager, eventCollector, requestMulti, serverEntityService, false, false);
+    ServerEntityRequest request = mockCreateEntityRequest();
+    TestingResponse response = mockResponse();
+    String config = "foo";
+    MessagePayload arg = mockCreatePayload(config);
+    invokeOnTransactionHandler(()->managedEntity.addRequestMessage(request, arg, null, response::complete, response::failure));
+    response.waitFor();
+    verify(response).complete(Mockito.any());
+    promote();
+
+    ServerEntityRequest request2 = mockDestroyEntityRequest();
+    TestingResponse response2 = mockResponse();
+    MessagePayload arg2 = mockInvokePayload();
+    invokeOnTransactionHandler(()->managedEntity.addRequestMessage(request2, arg2, null, response2::complete, response2::failure));
+    response2.waitFor();
+    verify(response2).failure(Mockito.any());
+  }
+  
+  @Test 
   public void testExecutionStrategy() throws Exception {
     // first create a passive entity and then promote
     ServerEntityRequest request = mockCreateEntityRequest();
@@ -802,6 +823,12 @@ public class ManagedEntityImplTest {
     return request;
   }
 
+  private ServerEntityRequest mockDestroyEntityRequest() {
+    ServerEntityRequest request = mockRequestForAction(ServerEntityAction.DESTROY_ENTITY);
+    when(request.getNodeID()).thenReturn(nodeID);
+    return request;
+  }
+  
   private ServerEntityRequest mockReconfigureEntityRequest() {
     ServerEntityRequest request = mockRequestForAction(ServerEntityAction.RECONFIGURE_ENTITY);
     when(request.getNodeID()).thenReturn(nodeID);

